File: codes/DSA 2023/balanced_paranthesis.py
#stack using list

def balanced_paranthesis(s): 
    stack =[]
    
    for i in s:

        if i == '(' or i =='{' or i == '[':
            stack.append(i)

        elif i == ')' :
            if len(stack) != 0 and stack[len(stack)-1] == '(':
                stack.pop()
            else : return False
            
        elif i == '}' :
            if len(stack) != 0 and stack[len(stack)-1] == '{':
                stack.pop()
            else : return False

        elif i == ']' :
            if len(stack) != 0 and stack[len(stack)-1] == '[':
                stack.pop()
            else : return False

    if len(stack) == 0:
        return True

print(balanced_paranthesis('[4*{3-(2+(5-1))}*(4-5)]'))

            
        

File: codes/DSA 2023/balanced_paranthesis.cpp
#include <iostream>
#include <stack>
using namespace std;

bool balanced_paranthesis(string s){
    stack <char> st;
    for(int i=0; i<s.length();i++){

        if(s[i]=='(')
        st.push(s[i]);

        else if(s[i]=='{')
        st.push(s[i]);

        else if(s[i]=='[')
        st.push(s[i]);

        else if(s[i]==')' ){
            if(!st.empty() && st.top()=='(')
            st.pop(); 
            else { return false;}
        }
        else if(s[i]=='}' ){
            if(!st.empty() && st.top()=='{')
            st.pop(); 
            else { return false;}
        }
        else if(s[i]==']' ){
            if(!st.empty() && st.top()=='[')
            st.pop(); 
            else { return false;}
        }
    }
    if(st.empty())
    return true;
}
int main(){
    string str = "[4*{3-(2+5-1)}*(4-5)]";
    if(balanced_paranthesis(str)){
        cout<<"balanced_paranthesis"<<endl;
    }
    else {cout<<"not balanced_paranthesis";}
     return 0;
}